[ItsMeNengYi] iP

ItsMeNengYi opened the pull request:
# Nen2
> "If you stop coding for 60 seconds, you will not be coding for 1 minute" - Me
Nen2 is free and it can 
- manage your tasks
* roasts you if you enter wrong command

All you need to do is,
1. download it from [here](https://github.com/ItsMeNengYi/ip/releases/tag/A-Jar).
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you

Ofcoz it's **FREE** 
> "No one want to pay for this" - Also Me

Who are our target users
- [x] YOU
- [x] your mom
- [ ] your dog
- [x] your dad

If you are a Java programmer, you can use it to practice Java too. Here's the main method:
```
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

Commit:
Rename chatbot Duke to Nen2

Nen2.java: Add greet(), exit() public static method and logo, seperator String field

Commit:
Nen2.java: Add processMessage() method to process input of user

Implement Echo feature

Commit:
Nen2.java: Add printList() method to print all the text input by user

Commit:
Create class Task

Implement ability to mark and unmark tasks

Nen2.java: Rename processMessage() to parseInput() for clearer naming

Commit:
Create classes Todo, Event and Deadline as subclass of Task

Task.java: Create of() factory method to produce respective tasks, changed Task class to abstract, implement findArgumentOf() to find arguments for each parameter, implement getDescription() to extract description of task from text

Task class is changed to abstract since it is no longer instantiated

Commit:
Write simple test to input.txt and EXPECTED.TXT to test UI and features such as todo, deadline, event, list, mark, unmark, and bye

Commit:
Create classes ArgumentMissingException, EmptyDescriptionException and InvalidInputException to handle respective errors

Commit:
Nen2: Create getIndex() to get the index of task needed to be manipulated

handle missing number error, index of task out of bound error and given argument is not number error

Commit:
Update javaDoc for classes and public methods

Commit:
Implement save and read data

Nen2.java: Add saveData() and loadSavedData()

Task.java: Add toData() to convert task into specific format of string and parseData() to parse data read from file into instance of object

Create FailToParseDataException which is thrown when data parsing fail in toData() method

Create a nen2.txt in data/ to keep track of user history

Commit:
Merge branch 'branch-Level-7'

Commit:
Implement Dates and Times recognition to nen2

Create DateTimeFormatIncorrectException class and is thrown when date time format is incorrect

Commit:
Merge branch 'branch-Level-8'

Commit:
Made Nen2 more OOP

Create Parser, Storage, Tasklist and Ui class to split the job originally in Nen2 class

Commit:
Add JavaDOC for public methods and classes

Commit:
Merge branch 'branch-A-MoreOOP'

Commit:
Put everything into a package name Nen2

Create a new Main class

Commit:
Merge branch 'branch-A-Packages'

Commit:
Add /data/ directory into .gitignore

Commit:
Merge branch 'add-gradle-support'

Commit:
Add gradle to Nen2

Commit:
Add two trivial tests for Parser and Task

Commit:
Implement find tasks by keyword feature

Commit:
Add JavaDoc to public methods and classes

Commit:
Merge branch 'branch-A-JavaDoc'

Commit:
Merge branch 'branch-Level-9'

Commit:
Configure and use checkStyle to detect coding style violations

Rename and sort packages into 4 categories: exceptions, nen2,tasks and utils

Commit:
Merge branch 'A-CheckStyle'

Commit:
Add GUI to Nen2 chatbot

Create Command class and multiple subtypes of Command class

Commit:
Merge branch 'branch-Level-10'

Commit:
Tweak code style

Jayden-Kim-NUS commented on src/main/java/Nen2/TaskList.java:
```diff
@@ -0,0 +1,92 @@
+        return listOfTasks.stream().map(Task::toData).toArray(String[]::new);
+    }
+
+    public Task[] findTaskWithKeyword(String keyword) {
```
Consider writing a javadocs header here to explain the method

Jayden-Kim-NUS commented on src/main/java/Nen2/Parser.java:
```diff
@@ -0,0 +1,111 @@
+                break;
+            case "find":
+                try
+                {
```
Suggested to use egyptian brackets for try catch block

Jayden-Kim-NUS commented on src/main/java/Nen2/Parser.java:
```diff
@@ -0,0 +1,111 @@
+                react += "Got it. I've added this task: \n" + t + "\n" +
+                        "Now you have " + taskList.size() +
+                        " tasks in the list." + "\n";
+            }
```
Would be good practice to put break; or //fallthrough even in default case

Jayden-Kim-NUS commented on src/main/java/Nen2/TaskList.java:
```diff
@@ -0,0 +1,92 @@
+
+    }
+
+    public TaskList(String[] arr) {
```
Perhaps add a javadocs header here to explain what this constructor does

Jayden-Kim-NUS reviewed (commented):
LGTM, slight coding standards issues but overall great.

ZShunRen commented on src/main/java/nen/commands/AddTaskCommand.java:
```diff
@@ -0,0 +1,45 @@
+import nen.utils.TaskList;
+
+/**
+ *  This class represents a end command which turn off Nen2
```
There is an extra whitespace character here

ZShunRen commented on src/main/java/nen/commands/AddTaskCommand.java:
```diff
@@ -0,0 +1,45 @@
+    }
+
+    /**
+     * Execute the command
```
Perhaps white spacing could be placed between the description and params/return in javadoc?

ZShunRen commented on src/main/java/nen/utils/Parser.java:
```diff
@@ -0,0 +1,74 @@
+            case "mark", "delete", "unmark":
+                arg = String.valueOf(getIndex(input));
+                break;
+            default:
+                action = input;
+                break;
```
Perhaps you can add cases for the task types here, e.g. "todo", "deadline", so that you can catch invalid commands here instead of having to catch it in the factory method of Tasks

ZShunRen commented on src/main/java/nen/commands/Command.java:
```diff
@@ -0,0 +1,61 @@
+ * @author Gan Ren Yick (A0276246X)
+ */
+public abstract class Command {
+    protected String name;
+    protected String description = "";
+    protected int arg;
```
I like the appropriate usage of access modifiers here

ZShunRen commented on src/main/java/nen/commands/DeleteCommand.java:
```diff
@@ -0,0 +1,32 @@
+     */
+    @Override
+    public void execute(TaskList taskList) {
+        description = "Noted. I've removed this task:\n"
+                + taskList.get(arg - 1).toString() + "\n";
+        taskList.get(arg - 1).markAsNotDone();
+        taskList.remove(arg - 1);
+        description += "Now you have " + taskList.size() + " tasks in the list." + "\n";
```
Perhaps marking the ask as not done is unnecessary, since you are about to remove it anyways?

ZShunRen reviewed (commented):
Overall very clean code that is well-documented. Apart from a few questions, I think it's very well done

Commit:
Move Main class inside package for JAR file generation

Commit:
Use assert feature to document important assumptions

Commit:
Refactor and improve the code quality

ItsMeNengYi committed:
Merge pull request #2 from ItsMeNengYi/branch-A-Assertions

Use assert feature to document important assumptions

Commit:
Merge branch 'branch-A-CodeQuality'

Commit:
Merge branch 'master' into branch-A-CodeQuality

Commit:
Add gradle.yml

Commit:
Merge branch 'branch-A-CI'

Commit:
Implement sort command which sort todo lexicografically, chronologically for event and deadline

User need to be able to sort the tasks

Create SortCommand class, implement sort(String) in TaskList.java, make Task implements Comparable<Task>, all subtype of Task class and Task class override compareTo(Task) method

Commit:
Merge branch 'branch-C-Sort'

Commit:
Fix nen2 is not saving progress and no response bug

Commit:
Add application title "Nen2"

Commit:
Add user guide to README.md and product screenshot

Commit:
Merge branch 'branch-A-UserGuide'

Commit:
Fix failed to load event data Issue

Commit:
Tweak Javadoc comments

ItsMeNengYi committed:
Update README.md

Commit:
Fix fail to create file issue

ItsMeNengYi committed:
Fix failed to create file issue

Commit:
Set launcher class as main class

ItsMeNengYi committed:
Merge pull request #5 from ItsMeNengYi/branch-A-CodeQuality

Merge branch 'master' into branch-A-CodeQuality
